### DISCUSSION TITLE
Drive drifting to the left in simulator

### DISCUSSION THREAD

**Discussion by tfield** [BroncBotz3481/YAGSL-Example#180] [Q&A] [answered]
We're using the latest version of YAGSL and it's looking great. We aren't using the controller/heading correction though, we're using our own.

However, we noticed the drive drifting to the left a bit with our field-oriented drive control, so we started running some experiments. We checked our absolute encoders and made some minor adjustments, but we still saw some drifting. 

We ended up in the simulator. We saw the drifting to the left using field-oriented drive. We wanted to rule out our heading correction so we switched to robot-oriented drive. We literally are calling SwerveDrive.drive(new ChassisSpeeds(val, 0, 0)), and the unit still drifts to the left in the simulator.

I tried tweaking absolute encoder offsets and it seems to change the drift characteristics a bit, but I feel like I'm guessing. It seems that one or two of the wheels aren't turning as fast as others... but they have the same configuration. I thought they could be rotated, but the sim shows them properly aligned.

Here is a video. I have a button that resets our pose to the end of the field, so I ran the test twice - once showing our telemetry and once showing the swerve modules.

Any ideas?

https://youtu.be/5m3gs3wsT-M

**Comment by thenetworkgrinch** [comment, marked as answer]
I updated it today which may address this.

**Reply by tfield** [reply]
I tried it - no joy - same behaviour. Is there something I can do to diagnose this?

Here's the drive code that's running - the point being it's passing a plain old `ChassisSpeeds(x,0,0)` into `swervelib.SwerveDrive.drive()`:

```
// in teleop drive command (for this test):
        swerve.driveRobotOriented(new ChassisSpeeds(velocity.getX(), 0, 0));

// in my main swerve drive class
    public final void driveRobotOriented(ChassisSpeeds velocity) {

        double x = velocity.vxMetersPerSecond;
        double y = velocity.vyMetersPerSecond;
        double w = velocity.omegaRadiansPerSecond;

        // Limit change in values. Note this may not scale evenly - one may reach desired
        // speed before another. This will be corrected the next time drive() is called.
// limiters disabled for this test
//        x = xLimiter.calculate(x);
//        y = yLimiter.calculate(y);
//        w = omegaLimiter.calculate(w);

        driveRawRobotOriented(new ChassisSpeeds(x, y, w));
    }


// implementation of driveRawRobotOriented

    @Override
    protected void driveRawRobotOriented(ChassisSpeeds velocity) {
        SmartDashboard.putString("Drive/Swerve/chassis", String.format("%.2f,%.2f m/s %.0f deg/s)",
            velocity.vxMetersPerSecond, velocity.vyMetersPerSecond,
            Rotation2d.fromRadians(velocity.omegaRadiansPerSecond).getDegrees()));

        SmartDashboard.putString("Drive/Swerve/velocity", String.format("%.2f m/s %.0f deg/s)",
            Math.hypot(velocity.vxMetersPerSecond, velocity.vyMetersPerSecond),
            Rotation2d.fromRadians(velocity.omegaRadiansPerSecond).getDegrees()));

        swerveDrive.drive(velocity, false, new Translation2d());
    }

```

**Reply by tfield** [reply]
Oh, scratch that - I just found [this in SwerveSubsystem](https://github.com/BroncBotz3481/YAGSL-Example/blob/5f20fedc7a417d2f8604d35e80869ffcaa8070ee/src/main/java/frc/robot/subsystems/swervedrive/SwerveSubsystem.java#L85C1-L86C1).

```     
swerveDrive.setCosineCompensator(!SwerveDriveTelemetry.isSimulation);
```

and it work like a charm - at least in the sim. I'll try it in real life on Tuesday.

Thanks!
